blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java

```diff
@@ -710,14 +711,14 @@ public BluePipelineStep getPipelineNodeStep(String id, Link parent) {
 
                 // New wrapper object based on current execution
                 FlowNodeWrapper newNodeWrapper = new FlowNodeWrapper(
-                    oldNodeWrapper.getNode(), // Use old node because we want to show the final id not intermediary
+                    currentNodeWrapper.getNode(),
```

Comment:
This is the main fix. The old comment doesn't make sense to me. My understanding is that in this case, we have a real node from the current Pipeline. Its ID will never change, even when the build completes, so we already have the "final" ID. Because the code checks for a match using `findNodeWrapperByParentageIn` if it doesn't find a direct ID match, the ID of the real node might not match the ID of the corresponding node from the old build. This leads to a bug when we use that ID in reference to the current build in other contexts where we expect it to be a stage/branch or whatever, since that ID might be something totally different in the current build.

Reply:
Maybe we should just delete `newNodeWrapper` and use `currentNodeWrapper` directly now that we aren't changing the fields?

Reply:
The root problem is the node ids *do* change during a run, thanks to the way this code walks the graph backwards instead of forwards.

Reply:
Changing this will fix some of the bad graphs, and break others that are currently working. I used the old node because it's from a *finished* run, which has the "final" node ids.

Reply:
> The root problem is the node ids do change during a run, thanks to the way this code walks the graph backwards instead of forwards.

The ID for an actual Pipeline `FlowNode` in a build will never change once it is added to the flow graph. Whether we walk the graph in reverse or not does not matter. In this particular loop, as far as I can tell, we are looping through `FlowNodeWrappers`s for `FlowNode`s that already exist in the flow graph of the current build, so their IDs will never change.

If you mean the problem is that a single `FlowNodeWrapper` that reference nodes from the current build can end up referencing a different node once the build progresses further because we are mutating `FlowNodeWrapper` or something is unstable in the visitor, that seems like a bug, and I am happy to look into fixing that here too if we can come up with a reproduction case.

> I used the old node because it's from a finished run, which has the "final" node ids.

The problem is that expecting the node IDs from a different build to match the node IDs in the current build is absolutely not something you can count on. It happens to work in some cases, but `findNodeByParentageIn` is not strict enough to guarantee that if two nodes match by that criteria their IDs also match, and nothing in a Pipeline will change the ID of a `FlowNode` once it has been added to the flow graph to make them match after the fact or something.

Further on in this code, where we use the IDs from the old build in what we return, is also a bit suspect, because one of the old IDs might refer to a different node in the current Pipeline, but since those nodes are marked as having not started it seems to not really matter. It would be safer to make the node IDs for the fake `FlowNodeWrapper`s some ID that could not possibly be a valid ID in the current Pipeline, but I don't know what assumptions the rest of the code makes about the IDs for the fake `FlowNodeWrapper`s.

Reply:
Yes I know the underlying canonical graph doesn't change, but due to the `PipelineNodeGraphVisitor` working backwards (and relies on events around structure rather than just the nodes themselves), it sometimes picks a different set of nodes to represent the same graph depending on where exactly the job is during a build. This is the underlying issue causing all of these problems, because the UI needs to be able to identify a node by id, both for navigation and selection, and (I believe) for updating the model during builds when SSE events come in. 

The drift of node ids causes the difficulty in merging the old and new graphs on the server-side, and also causes the UI state to not get correctly updated all the time, as well as user events not being tied to ids when the model held in the front-end happens to have the transiently-bad ids, the user clicks something, and then that id is no longer attached to that node by the time of the round-trip to the server looking for more info.

Reply:
I'm not trying to prescribe solutions, just trying to dump out my knowledge before it all bitrots away in my head.

Reply:
> it sometimes picks a different set of nodes to represent the same graph depending on where exactly the job is during a build

Ok, do you have any particular scenarios in mind that might cause this, so I can investigate them and hopefully at least add a test case for the future? 

One specific case with weird behavior that I am aware of is that if we have a sequential stage, the status computation for the `FlowNodeWrapper` for that stage will use the start of the sequential stage as the start (correct), and the start of the first stage inside of the sequential stage as the end (it should probably be the end of the sequential stage like for regular stages, but `StageChunkFinder` doesn't work for sequential stages). I don't think that would affect the stability of the listener the way you are describing (and it doesn't really matter because we don't display a status bubble for sequential stages), but it falls under the "weird behavior with PipelineNodeGraphVisitor" umbrella.

Reply:
Ok, I came up with two minimal test cases where nodes that show up while the build is in progress do not appear in the final graph. Both of them involve a `parallel` step with only one branch, so they're not particularly realistic, but they should help to figure out what is going on (and to check if my change affects the behavior or not).

Here is the Declarative test case:

```
pipeline {
  agent none
  stages {
    stage('outer') {
      parallel {
        stage('inner') {
          steps {
            sleep 5
          }
        }
      }
    }
  }
}
```

Here is the scripted test case:

```
stage('outer') {
  parallel inner: {
    stage('inner') {
        sleep 5
    }
  }
}
```

In both cases, `outer` is shown in a separate bubble preceding `inner` until the build completes, at which point only `inner` is shown.

Reply:
Ok, looking into those test cases, the main difference in the graph of `FlowNodeWrapper`s when the build is running vs completed is that when the build is running, the `FlowNodeWrapper.type` of the node for `inner` is `STAGE`, but when the build completes the `FlowNodeWrapper.type` for `inner` becomes `PARALLEL`. I'll keep looking to understand why and either fix it or at add an ignored test case, although I'll probably open a separate PR for that.

Reply:
I couldn't reproduce all of the situations that cause node drift reliably, it's definitely timing-sensitive. I think it occurs only when parallel stages are just starting or just ending, such that the nodes exist but not all the events for those nodes do yet. I don't really have more detail to give, other than my conclusion that it'd be quicker building a new graph visitor that uses the newer pipeline API via feature flag rather than trying to debug the exact event sequence that triggers it. Fresh eyes may be able to find it though.

Reply:
Another possibility might be adopting a different id scheme, that is tied to the structure of the graph (akin to an xpath or something) rather than just reusing the ids given to the pipeline nodes by Jenkins.